_posts/fastapi-rust-5-rocket-0.5.md

```diff
@@ -0,0 +1,212 @@
+
+## Step 2: Check and Check and Check
+
+I’ve now entered the stage where I’m repeatedly running `cargo check` and fixing the errors I’ve found. Unfortunately I came to a similar issue quite quickly to the one I faced when implementing my own `Handler` for `actix-web`: passing data to the `Future` doing the actual handling. The code changes for Step 2 can be found [in this commit](https://github.com/dbanty/rocket-lamb/commit/6a995b1b977675736fa32d96e1094ba3fcc05098).
+
+### Step 2.1: Solving Async References
+
+The `Fut` associated type on `Handler` requires that any lifetimes be `’static`, presumably so the `Future` can live longer than the `Handler` that spawned it (though I’m not sure why you’d want this). In any event, the `Handler` code I inherited from the existing `rocket_lamb` crate was passing `&mut self` to the now-asynchronous code which left me a bunch of errors saying something about needing `’static` lifetimes.
+
+At this point I was thoroughly lost, if I needed to pass data to something, and some trait outside of my control was requiring it to be `'static`, then I couldn’t pass references right? I didn’t want to clone all the data for every request, as doing so would be very inefficient. Even if I wanted to do that, some of the data I needed to pass was not `Clone`. I decided to go out on a limb and ask for help. I’ll admit I was afraid, as much as I try to contribute positively to the open source community, the responses I’ve received have been very mixed. However, luckily, one of the contributors to the `lamedh_` crates came through with the answer!
+
+I had heard of, though never used, `std::sync::Arc` before. I’m not really qualified to explain how this works, but I’ll do my best to convey _my_ understanding. `std::rc::Rc` effectively lets you have shared, read-only ownership of an object. If you do something like this:
+
+```rust
+use std::rc::Rc;
+
+let original = String::from("blah");
+let first = Rc::new(original);
+let second = Rc::clone(&first);
+```
+
+Then you effectively end up with two different owned values `first` and `second` which can be dereferenced to read data as if it were `original` (note that creating `first` took ownership of `original` so you can’t access that one anymore). Because the values are owned, you’re not passing references, and therefore don’t have to worry about lifetimes!
+
+This isn’t magic, `Rc` stands for “reference counting” which is a common way of keeping track of memory in more dynamic languages (e.g. Python). Behind the scenes, Rust is keeping track of how many times `Rc::clone` has been used for a given piece of data and once they’re _all_ dropped, it will free the memory. So even if `Handler` goes out of scope and gets dropped with a `Future` still on the loose, the data it was given is still perfectly valid!
```

Comment:
```suggestion
This isn’t magic, `Rc` stands for “reference counting” which is a common way of handling memory in more dynamic languages (e.g. Python). Behind the scenes, Rust is keeping track of how many times `Rc::clone` has been used for a given piece of data and once they’re _all_ dropped, it will free the memory. So even if `Handler` goes out of scope and gets dropped with a `Future` still on the loose, the data it was given is still perfectly valid!
```